src/pymor/algorithms/phdmd.py

```diff
@@ -0,0 +1,564 @@
+# This file is part of the pyMOR project (https://www.pymor.org).
+# Copyright pyMOR developers and contributors. All rights reserved.
+# License: BSD 2-Clause License (https://opensource.org/licenses/BSD-2-Clause)
+
+import numpy as np
+from scipy.linalg import eigh, lstsq, svd
+from scipy.sparse import diags
+from scipy.sparse.linalg import spsolve
+
+from pymor.algorithms.to_matrix import to_matrix
+from pymor.core.exceptions import PHDMDError
+from pymor.core.logger import getLogger
+from pymor.models.iosys import PHLTIModel
+from pymor.operators.block import BlockOperator
+from pymor.operators.constructions import VectorArrayOperator
+from pymor.operators.interface import Operator
+from pymor.operators.numpy import NumpyMatrixOperator
+from pymor.vectorarrays.block import BlockVectorSpace
+from pymor.vectorarrays.interface import VectorArray
+
+
+def phdmd(X, Y, U, Xdot=None, dt=None, H=None, initial_J=None, initial_R=None, initial_alpha=0.1, miniter=0,
+          maxiter=30, atol=1e-12, rtol=1e-10, weighted_rtol=1e-12, skew_procrustes_rtol=1e-12):
+    r"""Solve the pH DMD problem.
+
+    This method solves the optimization problem::
+
+        min ||Z - (J - R) T||_F
+
+    such that :math:`J` is a skew-symmetric matrix and :math:`R` is symmetric and positive
+    semidefinite. The data matrices :math:`T` and :math:`Z` are constructed as follows::
+
+        T = (X, U)^T, Z = (H \dot{X}, -Y)^T
+
+    from `X`, `Y`, `U`, and either derivative `Xdot` or an approximation of the derivative data
+    obtained from `X` itself.
+
+    The formal description of this algorithm can be found in :cite:`MMU22`.
+
+    Parameters
+    ----------
+    X
+        The |VectorArray| of the system state data.
+    Y
+        The |VectorArray| of the system output data.
+    U
+        The |VectorArray| of the system control data.
```

Comment:
At some point, we decided that inputs and outputs should be NumPy arrays, since they should be easy to export and import. Maybe it would be good to use NumPy arrays for Y and U (and convert them to `NumpyVectorArrays` internally if needed).